File: app/src/main/java/ru/art2000/calculator/settings/SettingsFragment.java
package ru.art2000.calculator.settings;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;

import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.SwitchPreferenceCompat;

import java.util.Objects;

import ru.art2000.calculator.BuildConfig;
import ru.art2000.calculator.R;

public class SettingsFragment extends PreferenceFragmentCompat {

    private Context mContext;
    private int dev = 0;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.settings);
        mContext = getActivity();

        /* tabList preference */
        ListPreference tabList = findPreference("tab_default");
        if (!getResources().getBoolean(R.bool.show_settings_as_default_tab)) {
            CharSequence[] tabEntries = tabList.getEntries();
            CharSequence[] tabValues = tabList.getEntryValues();
            CharSequence[] newTabEntries = new CharSequence[tabEntries.length - 1];
            CharSequence[] newTabValues = new CharSequence[tabValues.length - 1];
            short found = 0;
            for (int i = 0; i < tabValues.length; ++i) {
                CharSequence sequence = tabValues[i];
                if (sequence.equals("settings_tab")) {
                    found = 1;
                } else {
                    newTabEntries[i - found] = tabEntries[i];
                    newTabValues[i - found] = tabValues[i];
                }
            }
            tabList.setEntries(newTabEntries);
            tabList.setEntryValues(newTabValues);
        }
        tabList.setSummary(tabList.getEntry());
        tabList.setOnPreferenceChangeListener((preference, newValue) -> {
            tabList.setValue(newValue.toString());
            tabList.setSummary(tabList.getEntry());
            String stringValue = newValue.toString();
            PrefsHelper.setDefaultTab(stringValue);
            return true;
        });

        /* appTheme preference */
        ListPreference appTheme = findPreference("app_theme");
        appTheme.setSummary(appTheme.getEntry());
        appTheme.setOnPreferenceChangeListener((preference, newValue) -> {
            appTheme.setValue(newValue.toString());
            appTheme.setSummary(appTheme.getEntry());
            switch (newValue.toString()) {
                case "dark":
                    Objects.requireNonNull(getActivity()).getApplication().setTheme(R.style.AppTheme_Dark);
                    break;
                case "day_night":
                    Objects.requireNonNull(getActivity()).getApplication().setTheme(R.style.AppTheme_DayNight);
                    break;
                case "system":
                    Objects.requireNonNull(getActivity()).getApplication().setTheme(R.style.AppTheme_System);
                    Toast.makeText(mContext, R.string.daynight_support_message, Toast.LENGTH_LONG).show();
                    break;
                case "battery":
                    Objects.requireNonNull(getActivity()).getApplication().setTheme(R.style.AppTheme_Battery);
                    Toast.makeText(mContext, R.string.daynight_support_message, Toast.LENGTH_LONG).show();
                    break;
                default:
                    Objects.requireNonNull(getActivity()).getApplication().setTheme(R.style.AppTheme);
                    break;
            }
            PrefsHelper.setAppTheme(newValue.toString());
            Intent intent = getActivity().getIntent();
            intent.setAction("ru.art2000.calculator.action.SETTINGS");
            Activity parent = getActivity();
            parent.finish();
            parent.startActivity(intent);
            getActivity().recreate();
            return true;
        });

        /* saveCurrencyConversion preference */
        SwitchPreferenceCompat saveCurrencyConversion = findPreference("save_currency_value");
        saveCurrencyConversion.setOnPreferenceChangeListener((preference, newValue) -> {
            PrefsHelper.setShouldSaveCurrencyConversion((Boolean) newValue);
            return true;
        });

        /* zeroDiv preference */
        SwitchPreferenceCompat zeroDiv = findPreference("zero_div");
        zeroDiv.setOnPreferenceChangeListener((preference, newValue) -> {
            PrefsHelper.setZeroDivResult((Boolean) newValue);
            return true;
        });

        /* unitView preference */
        ListPreference unitView = findPreference("unit_view");
        unitView.setSummary(unitView.getEntry());
        unitView.setOnPreferenceChangeListener((preference, newValue) -> {
            unitView.setValue(newValue.toString());
            unitView.setSummary(unitView.getEntry());
            PrefsHelper.setUnitViewType(newValue.toString());
            PrefsHelper.setUnitViewChanged();
            return true;
        });

        /* appVersion preference */
        Preference appVersion = findPreference("app_version");
        appVersion.setSummary(BuildConfig.VERSION_NAME);
        appVersion.setOnPreferenceClickListener(preference -> {
            dev++;
            if (dev == 1) {
                dev = 0;
                startActivity(new Intent(mContext, InfoActivity.class));
            }
            return false;
        });
    }
}

File: settings/src/androidMain/kotlin/ru/art2000/calculator/settings/view/SettingsFragment.kt
package ru.art2000.calculator.settings.view

import android.content.Intent
import android.content.pm.PackageManager.PackageInfoFlags
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.view.children
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.SwitchPreferenceCompat
import androidx.recyclerview.widget.RecyclerView
import dagger.hilt.android.AndroidEntryPoint
import ru.art2000.calculator.common.preferences.SettingsSetup
import ru.art2000.calculator.common.model.MainPage
import ru.art2000.calculator.common.preferences.MainTabData
import ru.art2000.calculator.common.view.MainScreenPreferenceFragment
import ru.art2000.calculator.settings.BuildConfig
import ru.art2000.calculator.settings.R
import ru.art2000.calculator.settings.preferences.DYNAMIC_COLORS_AVAILABLE
import ru.art2000.calculator.settings.preferences.PreferenceKeys
import ru.art2000.calculator.settings.preferences.PreferenceKeys.KEY_APP_THEME
import ru.art2000.calculator.settings.preferences.PreferenceKeys.KEY_AUTO_DARK_THEME
import ru.art2000.calculator.settings.preferences.PreferenceKeys.KEY_DARK_THEME_ACTIVATION
import ru.art2000.calculator.settings.preferences.PreferenceKeys.KEY_DARK_THEME_DEACTIVATION
import ru.art2000.calculator.settings.preferences.PreferenceKeys.KEY_DYNAMIC_COLORS
import ru.art2000.extensions.preferences.AppTheme
import ru.art2000.extensions.views.isLandscape
import java.text.DateFormat
import java.util.*
import javax.inject.Inject

@AndroidEntryPoint
internal class SettingsFragment : MainScreenPreferenceFragment() {

    private var dev = 0

    companion object {
        private val CLICKS_TO_OPEN_HIDDEN_INFO = if (BuildConfig.DEBUG) 1 else 7

        private const val KEY_APP_VERSION = "app_version"

        private val AUTO_THEMES_ANDROIDX = listOf(
            AppTheme.SYSTEM, AppTheme.BATTERY,
        ).map { it.name.lowercase() }

        private val DAY_NIGHT_THEME_KEY = AppTheme.DAY_NIGHT.name.lowercase()

        private val AUTO_THEMES = AUTO_THEMES_ANDROIDX + DAY_NIGHT_THEME_KEY
    }

    @Inject
    lateinit var settingsSetups: Map<MainPage, @JvmSuppressWildcards SettingsSetup>

    @Inject
    lateinit var tabDatas: Map<MainPage, @JvmSuppressWildcards MainTabData<*>>

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.general_settings)

        settingsSetups
            .toSortedMap(Comparator.comparingInt { it.ordinal })
            .values
            .forEach {
                with (it) {
                    populateSettings()
                }
            }

        addPreferencesFromResource(R.xml.info)

        val dynamicColors = findPreference<SwitchPreferenceCompat>(KEY_DYNAMIC_COLORS)
        /* appTheme preference */
        val appTheme = findPreference<ListPreference>(KEY_APP_THEME)
        /* appAutoDarkTheme preference */
        val appAutoDarkTheme = findPreference<SwitchPreferenceCompat>(KEY_AUTO_DARK_THEME)
        /* appAutoDarkTheme preference */
        val appDarkThemeActivationTime =
            findPreference<Preference>(KEY_DARK_THEME_ACTIVATION)
        /* appAutoDarkTheme preference */
        val appDarkThemeDeactivationTime =
            findPreference<Preference>(KEY_DARK_THEME_DEACTIVATION)

        val dynamicColorsDisable = listOf(
            appTheme, appAutoDarkTheme,
            appDarkThemeActivationTime, appDarkThemeDeactivationTime,
        )

        val themingGroup = dynamicColors?.parent

        fun areDynamicColorsEnabled() = DYNAMIC_COLORS_AVAILABLE
                && (dynamicColors?.isChecked ?: false)

        fun showThemingPreferences(dynamicColorsEnabled: Boolean) {
            if (dynamicColorsEnabled) {
                dynamicColorsDisable.forEach { preference ->
                    preference?.also { themingGroup?.removePreference(it) }
                }
            } else {
                dynamicColorsDisable.forEach { preference ->
                    preference?.also { themingGroup?.addPreference(it) }
                }
            }
        }

        showThemingPreferences(areDynamicColorsEnabled())

        dynamicColors?.also {
            if (!DYNAMIC_COLORS_AVAILABLE) {
                it.parent?.removePreference(it)
            }
        }

        appTheme?.also {
            appTheme.onPreferenceChangeListener =
                Preference.OnPreferenceChangeListener { _, newValue: Any ->
                    if (appTheme.value == newValue) {
                        return@OnPreferenceChangeListener false
                    }
                    appTheme.value = newValue.toString()
                    if (newValue.toString() in AUTO_THEMES_ANDROIDX) {
                        Toast.makeText(
                            requireContext(),
                            R.string.daynight_support_message,
                            Toast.LENGTH_LONG
                        ).show()
                    }
                    true
                }
        }

        appAutoDarkTheme?.also {
            if (appTheme?.value !in AUTO_THEMES) {
                it.parent?.removePreference(it)
                return@also
            }

            it.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
                appAutoDarkTheme.isChecked = newValue as Boolean
                true
            }
        }

        appDarkThemeActivationTime?.also {
            if (appTheme?.value != DAY_NIGHT_THEME_KEY) {
                it.parent?.removePreference(it)
                return@also
            }
        }

        appDarkThemeDeactivationTime?.also {
            if (appTheme?.value != DAY_NIGHT_THEME_KEY) {
                it.parent?.removePreference(it)
                return@also
            }
        }

        val tabPreference = findPreference<ListPreference>(PreferenceKeys.KEY_TAB_DEFAULT)
        tabPreference?.also {
            val data = tabDatas
                .filterKeys { page -> BuildConfig.DEBUG || page != MainPage.SETTINGS }
                .toSortedMap(Comparator.comparingInt { page -> page.ordinal })
                .values

            it.entries = data.map<MainTabData<*>, CharSequence> {
                    item -> requireContext().getString(item.titleRes)
            }.toTypedArray() + it.entries
            it.entryValues = data.map { item -> item.key }.toTypedArray() + it.entryValues
        }

        /* appVersion preference */
        val appVersion = findPreference<Preference>(KEY_APP_VERSION)
        val buildDate = DateFormat
            .getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM)
            .format(Date(BuildConfig.BUILD_TIME))
        val suffix = if (BuildConfig.DEBUG) "-debug" else ""
        val versionName = requireContext().packageManager.let {

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                it.getPackageInfo(
                    requireContext().packageName, PackageInfoFlags.of(0)
                )
            } else {
                it.getPackageInfo(
                    requireContext().packageName, 0
                )
            }

        }.versionName
        appVersion?.summary =
            "$versionName$suffix ($buildDate)"
        appVersion?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            dev++
            if (dev == CLICKS_TO_OPEN_HIDDEN_INFO) {
                dev = 0
                startActivity(Intent(requireContext(), InfoActivity::class.java))
            }
            false
        }
    }

    override val topViews: List<View>
        get() = listOf(findRecyclerView(view as ViewGroup)!!)

    override val bottomViews: List<View>
        get() = if (requireContext().isLandscape) topViews else emptyList()

    private fun findRecyclerView(group: ViewGroup): RecyclerView? {
        for (v in group.children) {
            if (v is RecyclerView) return v
            if (v is ViewGroup) {
                val res = findRecyclerView(v)
                if (res != null) return res
            }
        }

        return null
    }
}